From f40ac3b771794dda7129066b2fd89c01d698c92f Mon Sep 17 00:00:00 2001
From: Sean McGrail <549813+skmcgrail@users.noreply.github.com>
Date: Thu, 29 Jan 2026 16:24:03 -0800
Subject: [PATCH] 1-byte OOB read in EVP_PKEY_asn1_find_str length calculation

We would like to thank Joshua Rogers (https://joshua.hu/) of AISLE Research Team (https://aisle.com/) for reporting this issue.

(cherry picked from commit 921c6465918e2d118533199e6e2959f82483dde5)
---
 crypto/evp_extra/evp_asn1.c  | 11 +++++------
 crypto/evp_extra/evp_test.cc | 11 +++++++++++
 crypto/evp_extra/internal.h  |  2 ++
 3 files changed, 18 insertions(+), 6 deletions(-)

diff --git a/crypto/evp_extra/evp_asn1.c b/crypto/evp_extra/evp_asn1.c
index e394494cb..b3a7819db 100644
--- a/crypto/evp_extra/evp_asn1.c
+++ b/crypto/evp_extra/evp_asn1.c
@@ -612,15 +612,14 @@ const EVP_PKEY_ASN1_METHOD *EVP_PKEY_asn1_find_str(ENGINE **_pe,
   for (size_t i = 0; i < (size_t)EVP_PKEY_asn1_get_count(); i++) {
     const EVP_PKEY_ASN1_METHOD *ameth = EVP_PKEY_asn1_get0(i);
 
-    const size_t longest_pem_str_len = 10;  // "DILITHIUM3"
-
     const size_t pem_str_len =
-        OPENSSL_strnlen(ameth->pem_str, longest_pem_str_len);
+        OPENSSL_strnlen(ameth->pem_str, MAX_PEM_STR_LEN);
 
     // OPENSSL_strncasecmp(a, b, n) compares up to index n-1
-    const size_t cmp_len =
-        1 + ((name_len < pem_str_len) ? name_len : pem_str_len);
-    if (0 == OPENSSL_strncasecmp(ameth->pem_str, name, cmp_len)) {
+    if (name_len != pem_str_len) {
+      continue;
+    }
+    if (0 == OPENSSL_strncasecmp(ameth->pem_str, name, name_len)) {
       return ameth;
     }
   }
diff --git a/crypto/evp_extra/evp_test.cc b/crypto/evp_extra/evp_test.cc
index f98bf73a7..4874a20c4 100644
--- a/crypto/evp_extra/evp_test.cc
+++ b/crypto/evp_extra/evp_test.cc
@@ -61,6 +61,7 @@
 #include <string.h>
 
 #include "../fipsmodule/evp/internal.h"
+#include "internal.h"
 
 OPENSSL_MSVC_PRAGMA(warning(push))
 OPENSSL_MSVC_PRAGMA(warning(disable: 4702))
@@ -1528,3 +1529,13 @@ TEST(EVPTest, PKEY_asn1_find_str) {
   ASSERT_FALSE(ameth);
   ASSERT_FALSE(EVP_PKEY_asn1_get0_info(&pkey_id, &pkey_base_id, &pkey_flags, &pinfo, &pem_str, ameth));
 }
+
+TEST(EVPTest, ASN1MethodCheckPemStrLengthInvariant) {
+  for (int i = 0; i < EVP_PKEY_asn1_get_count(); i++) {
+    SCOPED_TRACE(i);
+    const EVP_PKEY_ASN1_METHOD *method = EVP_PKEY_asn1_get0(i);
+    ASSERT_NE(method, nullptr);
+    ASSERT_NE(method->pem_str, nullptr);
+    EXPECT_LE(OPENSSL_strnlen(method->pem_str, strlen(method->pem_str)+1), MAX_PEM_STR_LEN);
+  }
+}
diff --git a/crypto/evp_extra/internal.h b/crypto/evp_extra/internal.h
index 85b4f123d..d04fe3ad3 100644
--- a/crypto/evp_extra/internal.h
+++ b/crypto/evp_extra/internal.h
@@ -13,6 +13,8 @@
 #define PKCS8_VERSION_TWO 1
 #define ED25519_PUBLIC_KEY_OFFSET 32
 
+#define MAX_PEM_STR_LEN ((size_t)10) // "ED25519ph"
+
 typedef struct {
   uint8_t pub[32];
   uint8_t priv[32];
